### PR TITLE
BFD-1676: Introduce parallel performance regression test suite

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-1676](https://jira.cms.gov/browse/BFD-1676)

**User Story or Bug Summary:**

As a BFD engineer, I want to know if my new changes have significantly impacted the performance of the system so that I can avoid slowing down the endpoint responses.

---

### What Does This PR Do?

This PR:  

* Adds two new Python load test suites/files -- `v1/baseline_tests.py` and `v2/baseline_tests.py`
  * These files include all of the single tests (excluding those from `v2/partially_adjudicated.py` for `v2/baseline_tests.py`) from their respective API versions. However, when running one of these two new baseline files these tests run in parallel (with equal weighting to each endpoint)
  * Some work is done to attempt to bypass caching by setting `Cache-Control` headers and by shuffling input data per Locust `User`
  * Some expected SLA values were added to `common/validation.py` for both V1 and V2 baseline tests to ensure SLAs for these tests
* Refactors the way YAML configuration data is loaded
  * YAML configuration is read and merged with CLI arguments with the CLI taking priority, allowing users to re-use values in existing configuration files
  * Users can now specify a path to a _readonly_ (unless the file is `./config.yml`) YAML configuration file 
    * I expect this to be useful for pipeline integration, as the pipeline could provide connection details via the CLI, but could read users, spawn rate, and more via `configPath` without them being hardcoded in a `Jenkinsfile` or other source code
    * If the path is not `config.yml`, then the file is only read from and _not_ written to (the merged values are written to `config.yml`). I felt this behavior was preferable, especially if a developer is creating and testing a particular configuration as they would otherwise have to discard changes to the config file between runs
  * > **Note:** These changes to the configuration reading were made so that values that might vary depending on the target machine (such as user count, spawn rate, runtime, etc.) would not need to "hardcoded" in a `Jenkinsfile` or recorded in an external document  
* Adds YAML configuration files for both V1 and V2 baseline tests running against a `c5xlarge` target under the `config/` directory
  * Running these tests on a test instance would look something like `python3 runtests.py --configPath="config/config.baseline_v1_c5xlarge.yml --testFile="v1/baseline_tests.py" ...`
* Modifies the `README` to include some examples of how to run the tests with the configuration refactors, and also to fix some inaccuracies related to database CLI arguments

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Are these new test suites representative of what we feel is "baseline"? Are these suites suitable for identifying performance regressions?
* Is any additional documentation required?
* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/ccs-ops-misc/load_test/README.md [hunk 1]
```diff
@@ -70,24 +70,34 @@ The tests are run with parameters that specify all the test params and test file
...
 **--homePath** : (Required) : The path to your home directory on the local box, such as /home/logan.mitchell/
 
 **--clientCertPath** : (Required) : The path to the PEM file we copied/modified earlier. Should be located in your home directory like this: ~/bluebutton-backend-test-data-server-client-test-keypair.pem
 
-**--databaseHost** : (Required) : The database host, used to pull data. Get this from the database endpoint we noted in the earlier step "Getting The Database Instance"
-
-**--databaseUsername** : (Required) : The username for connecting to the database, can be found in Keybase. Make sure to use the correct one for the environment you're connecting to.
-
-**--databasePassword** : (Required) : The password for connecting to the database, can be found in Keybase. Make sure to use the correct one for the environment you're connecting to.
+**--databaseUri** : (Required) : The URI used for connecting to the database server. Needs to include username, password (both can be found in Keybase, make sure to use the correct one for the environment you're connecting to), hostname and port.
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/README.md:93]: Thanks for fixing this!

File: ops/ccs-ops-misc/load_test/README.md [hunk 2]
```diff
@@ -70,24 +70,34 @@ The tests are run with parameters that specify all the test params and test file
...
 
 **--testHost** : (Required) : The load balancer or single node to run the tests against. The environment used here should match the same environment (test, prod-sbx, prod) as the database, so we pull the appropriate data for the environment tested against. Note that when using a single node, you should specity the Ip AND the port for the application.
 
 **--testFile** : (Required) : The path to the test file we want to run.
 
+**--configPath** : (Optional) : The path to a YAML configuration file that will be read from (but not _written_ to) for the values specified here. The values in this configuration file will be merged with values from the CLI, with the CLI values taking priority. The resulting merged values will then be written to the root `config.yml` file. If not provided, defaults to `config.yml` (the root YAML configuration file).
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/README.md:99]: A little specific but it may be written to, if you use the root path or use the default, right? We read from this which may also be the place we write to. Not sure if we can reword this to be more specific, or if its worth doing so, since its relatively clear as is.

File: ops/ccs-ops-misc/load_test/common/url_path.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import time
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/common/url_path.py:1]: Is this being used?

File: ops/ccs-ops-misc/load_test/common/url_path.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+from typing import Dict
+from urllib.parse import urlencode
+
+'''
+Creates a query path from a base path (i.e. /v2/fhir/Coverage) and a dictionary of query parameters
+'''
+def create_url_path(path: str, query_params: Dict[str, str] = {}) -> str:
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/common/url_path.py:8]: Huh didnt know you could specify types for params in python, I like it.

File: ops/ccs-ops-misc/load_test/common/validation.py [hunk 5]
```diff
@@ -9,13 +9,21 @@
...
     SLA_PATIENT : [1000, 3000, 5000, 8000],
     SLA_EOB_WITH_SINCE : [100, 250, 1000, 3000],
-    SLA_EOB_WITHOUT_SINCE : [500, 1000, 3000, 6000]
+    SLA_EOB_WITHOUT_SINCE : [500, 1000, 3000, 6000],
+    # The following values were selected for 5 concurrent users against
+    # a target c5xlarge AWS instance running bfd-server 
+    SLA_V1_BASELINE : [10, 325, 550, 10000],
+    # The following values were selected for 3 concurrent users against
+    # a target c5xlarge AWS instance running bfd-server
+    SLA_V2_BASELINE : [10, 325, 550, 10000],
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/common/validation.py:26]: these will eventually be pulled from the new relic recorded average or whatever we use for metrics; maybe add a TODO here

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+server_public_key = setup.loadServerPublicKey()
+setup.disable_no_cert_warnings(server_public_key, urllib3)
+setup.set_locust_env(config.load())
+
+mbis = data.load_mbis()
+eob_ids = data.load_bene_ids()
+last_updated = data.get_last_updated()
+cursor_list = data.load_cursors("v1")
+
+class BFDUser(HttpUser):
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/v1/baseline_tests.py:22]: Could you add some documentation for this class please?

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
+import random
```
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/v1/baseline_tests.py:1]: Maybe we want to call this the regression suite or something instead of baseline tests? The baseline will be used to validate them, but they are intended to be used as regression tests. (same with the v2 name)

**Review by lsmitchell — COMMENTED** [review]
Great work! A couple minor things, but this is looking really nice.

File: ops/ccs-ops-misc/load_test/common/url_path.py (hunk 3 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/url_path.py:1]: Good catch! I must have accidentally imported that automatically when writing this, I'll remove it.

File: ops/ccs-ops-misc/load_test/common/url_path.py (hunk 4 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/url_path.py:7]: Yeah, I didn't either until working on this. Albeit I think they're just type hints opposed to actual compiler enforced types, so I think a caller could still pass a non-`str` type for `path` or a `list` for `query_params` and you there'd be no complaints from Python until runtime.

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        copied_cursor_list = cursor_list.copy()
+        random.shuffle(copied_cursor_list)
+
+        self.eob_ids = copied_eob_ids
+        self.mbis = copied_mbis
+        self.cursor_list = copied_cursor_list
+
+    def get_eob(self) -> int:
+      if len(self.eob_ids) == 0:
+            errors.no_data_stop_test(self)
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py:39]: This is a good point; for instance we have many fewer contract data pieces than generic bene ids; this means the test may end earlier than we want if we equally weight all tests.

[the remark this answers is not in the document]
Comment (andrew-kramer) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py:39]: Also, I think we probably consume those ids faster for some tests than others. There are like five or six that use bene_ids, a couple that use contract ids, and a couple more with MBIs. It may also depend on how quickly the requests come back, since we're firing all of them as quickly as we can.

[the remark this answers is not in the document]
Comment (malessi) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py:39]: This is a good point, and I didn't really consider that the list of contract URLs is less than the list of `bene_id`s, so in a prolonged test run of the suite we may run out of contract URLs to test against. As it stands, I was seeing approximately 25 requests per second per concurrent user for V2 and about double that for V1, so to your point @andrew-kramer over a 3 minute run we end up only sending about 7500 to 10000 requests _total_.

Note that each Locust `User` _copies_ the initial list of each type of input data (MBIs, `bene_id`s and contract URLs) in its `on_start(...)`. This means each concurrent user has their own set of independent data, by the way.

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+import common.validation as validation
+from locust import HttpUser, task, events, tag
+
+client_cert = setup.getClientCert()
+server_public_key = setup.loadServerPublicKey()
+setup.disable_no_cert_warnings(server_public_key, urllib3)
+setup.set_locust_env(config.load())
+
+mbis = data.load_mbis()
+eob_ids = data.load_bene_ids()
```
[the remark this answers is not in the document]
Comment (malessi) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py:18]: Yeah, I'm definitely in favor of that. I'll go ahead and rename all `eob_ids` to `bene_ids` and also rename `get_eob_id()` to `get_bene_id()`, too.

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py (hunk 7 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py:1]: Yeah, I think that's a better name for these suites. I'll rename them to the following:

* `v1/baseline_tests.py` -> `v2/regression_suite.py`
* `v2/baseline_tests.py` -> `v2/regression_suite.py`

File: ops/ccs-ops-misc/load_test/common/validation.py (hunk 5 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/validation.py:26]: Sounds good, I wanted to add some context as to why I chose those SLAs. I'll add a `TODO` above the dictionary here stating that we'll eventually be pulling these numbers from metrics.

File: ops/ccs-ops-misc/load_test/README.md (hunk 2 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/README.md:99]: Yep, you're correct. If `configPath` is `./config.yml` (the default config path, or the root configuration file) we will read _and_ write to it. If the `configPath` is _not_ `./config.yml` (some other YAML file) then we will _read_ from `configPath` and then write to `./config.yml` so that the file at `configPath` isn't overwritten unexpectedly.

I think I'll take a variation of what I wrote above here and rewrite that explanation, because looking at it now it's definitely a bit confusing.

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py (hunk 6 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py]: Yeah, I'll add some documentation to both regression suites and to their class-level methods.

File: ops/ccs-ops-misc/load_test/runtests.py [hunk 10]
```diff
@@ -91,48 +99,60 @@ class configData: pass
...
             workerThreads = arg
         elif opt == "--resetStats":
-            configData.resetStatsAfterClientSpawn = True
+            configData["resetStatsAfterClientSpawn"] = True
         else:
             print(helpString)
             sys.exit()
 
+    ## Read the specified configuration file and discard any falsey key
+    storedConfigData = {k: v for k, v in config.load_from_path(configData["configPath"]).items() if v}
```
Comment (andrew-kramer) [inline ops/ccs-ops-misc/load_test/runtests.py:130]: I get a KeyError here if I run without a configPath set, but configPath isn't listed as a required field. We should do one or the other -- my opinion is, it should probably not be required.

Comment (malessi) [reply ops/ccs-ops-misc/load_test/runtests.py]: Good catch! That's definitely a mistake on my end. I intended it to be optional, I just didn't catch that I was reading from `configData` prior to `configData["configPath"]` being set by merging with `defaultConfigData`. 

I've fixed this so that we use `Dict.get(...)` to fallback to `defaultConfigData["configPath"]` if `--configPath` was not passed-in. I've tested this locally and `--configPath` is now optional as expected.

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrew-kramer) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py]: Important note: in my own testing, if you run a Locust Task and it never makes a request to the API (i.e. you return immediately or after discovering that there's no data), then it doesn't count that run in the statistics. So it would be ok to do this:

```python
@task
def some_test(self):
    if self.bene_ids:
        # There is some data in self.bene_ids, so run the test
        self.client.get("url", params={ "_id": self.bene_ids.pop() })
    else:
        # There is no data, so just return and this test won't count
        return
```

If there is a bene_id there, it will run and count the test. If not, then the `if` will fail and the test will return without calling `client.get`. Locust will run other tests for which there is data until it runs out of that, too.

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py (hunk 6 above)
Comment (malessi) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py]: @lsmitchell I've added documentation comments to both files; let me know if you were thinking of adding external documentation or if you think that the file comments I've added are sufficient.

**Review by lsmitchell — DISMISSED** [review]
Nice work!

File: ops/ccs-ops-misc/load_test/v1/baseline_tests.py (hunk 6 above)
Comment (lsmitchell) [reply ops/ccs-ops-misc/load_test/v1/baseline_tests.py]: That should be fine, just wanted some small explanation of what the tests are doing in the file.

**Comment by christrotteradhoc** [comment]
This looks really great! Nice improvements here. Do you have any data on the stability of the tests from one run to the next that can help inform how much variation is normal? I think a section in the README or somewhere else that explains the steps and gives guidance on evaluating the results of a performance regression run would be really helpful.

**Comment by malessi** [comment]
> 

Yep, I do have some data (specifically for the V2 suite, but I have no reason to believe that the V1 suite would be any less consistent). [Here's a Google Sheet](https://docs.google.com/spreadsheets/d/1ZWFTwaN7TXDRxqqZ_dM_8ZRTl2uIS6337S7xIwaynms/edit?usp=sharing)
 with the statistics from 3 back-to-back test runs against the same target and with the same values (users, spawn rate, runtime). 

I've created charts comparing responses/second, 50%s and total average response times across the 3 runs. It's apparent that the results are _very_ consistent once the target machine is warmed-up (runs 2 and 3 have near-identical data) whereas a "cold" run tends to be slower in the beginning.

Here's the charts (I don't know why the first run has less data-points, the run didn't fail and it ran for the same amount of time as runs 2 and 3):

Combined Responses_Second for V2 Regression Suite Against c5xlarge Target

Combined 50%s for V2 Regression Suite Against c5xlarge Target

Combined Total Avg Response for V2 Regression Suite Against c5xlarge Target

**Comment by malessi** [comment]
> You might want to switch the data acquisition sql back to TABLE_SAMPLE and see the results; the holdover was to pull a static number of records from the DB which means your second and third tests might be using cached data. (We want to use table sampled results anyway, and I think the thing that was preventing us from doing so has been fixed now)

That's a good idea, the results definitely seem like some form of caching on the target machine and/or database. I'll use `TABLESAMPLE` for as much of the data as I can and see if anything changes.

**Comment by christrotteradhoc** [comment]
@malessi thanks for providing that data. The information that I am referring to is guidance on how much variation is expected from run to run for the purpose of correctly evaluating the absence of regression. However if this is going to be included in another story then that works for me.

**Comment by malessi** [comment]
> @malessi thanks for providing that data. The information that I am referring to is guidance on how much variation is expected from run to run for the purpose of correctly evaluating the absence of regression. However if this is going to be included in another story then that works for me.

Gotcha -- given now that we may not want to use `TAMPLESAMPLE`ing when running the regression suite (whereas we'd probably want it for the _load tests_) I'd say that those charts are fairly representative of the variation expected between subsequent runs using _"known" input data_ (same bene IDs/MBIs/contact URLs from run-to-run). 

However, I'd like to gather more data and have a method of storing and loading said data before adding any documentation about variance here -- at least until [BFD-1688](https://jira.cms.gov/browse/BFD-1688) is done (more likely whenever we actually implement the PoC from 1688).

**Review by lsmitchell — APPROVED** [review]

**Review by andrew-kramer — APPROVED** [review]
Looks great! Thanks, Mitch!

**Review by mackec1 — APPROVED** [review]
The only thing that I wondered about was why we have two distinct code paths for version (v1 vs. v2) when the v[1|2] is the only diffference in the request URL.

**Comment by malessi** [comment]
> 

This is going to be _mostly_ resolved by the changes in [BFD-1655](https://github.com/CMSgov/beneficiary-fhir-data/pull/1083).

**Comment by andrew-kramer** [comment]
As Mitch mentioned, it's mostly resolved in BFD-1655. Also, there are a few important differences - at least one endpoint uses the v1/Patient/{id} format, whereas there's no test for that endpoint in v2.

**Review by christrotteradhoc — APPROVED** [review]
